File: SEM2/ALGO/quicksort.cpp
#include<iostream>
#include<vector>
using namespace std;
void displayArray(vector<int> &array){
    cout<<"Displaying Your Array..."<<endl;
    for (int i = 0; i < array.size(); i++){
        cout<<array[i]<<" ";
    }
    cout<<endl;
}
void swap(vector<int> &array,int a,int b){
    int temp = array[a];
    array[a]  = array[b];
    array[b]  = temp;
}

int partition(vector<int> &array,int start,int end){
    int index = start-1;
    int pivot = end;
    while (start<end){
        if (array[start]<=array[pivot]){
            index+=1;
            swap(array,index,start);
        }
        start+=1;
    }
    index+=1;
    swap(array,index,pivot);
    return index;
}
void quickSort(vector<int> &array,int start,int end){
    if (start<end){
        int partion = partition(array,start,end);
        quickSort(array,start,partion-1);
        quickSort(array,partion+1,end);
    }
}

int main(){
    vector<int> array = {10,9,8,7,6,5,4,3,2,1};
    
    displayArray(array);
    quickSort(array,0,array.size()-1);
    displayArray(array);

    return 0;
}

File: SEM2/ALGO/ASSIGNMENT 2/quicksort.py
import random
import time
class QuickSort:
    def __init__(self,size) -> None:
        self.arr = self.generateRandomArrayOfSize(size)
        self.count = 0
    def generateRandomArrayOfSize(self,x):
        res = []
        for i in range(x):
            res.append(random.randint(1,100000))
        return res
    def getComparisonCount(self):
        return self.count
    def measure_execution_time(self):
        start_time = time.time()
        self.sort()
        end_time = time.time()
        return end_time - start_time
    def partition(self, array, start, end):
        index = start - 1
        pivot = end
        for i in range(start, end):
            if array[i] < array[pivot]:
                index += 1
                array[index], array[i] = array[i], array[index]
            self.count+=1
        index += 1
        array[index], array[pivot] = array[pivot], array[index]
        self.count+=1
        return index

    def quickSort(self, array, start, end):
        if start < end:
            p = self.partition(array, start, end)
            self.quickSort(array, start, p - 1)
            self.quickSort(array, p + 1, end)
    def sort(self):
        self.quickSort(self.arr,0,len(self.arr)-1)
